### ISSUE TITLE
Multi-input/output neural networks

### ISSUE THREAD

**Issue by pvdhove** [owlbarn/owl#307]
1. At the moment, it is not possible to get an output for the following neural network:
```
let input1 = input [|n; m|] in
let input2 = input [|n; k|] in
let nn =
  concatenate ~axis:1 [|input1; input2|]
  |> get_network
```
because `input1` and `input2` don't have a common ancestor.

2. It is not possible to get both outputs using only `get_network` and `model` for
```
let input = input [|n; m|] in
let output1 = activation ... input1 in
let output2 = activation ... input1 in ...
```
without creating two different networks, and thus computing twice the common parts of both networks.

For the inference mode, it is possible to run a DFS from the outputs using the backward edges of the network (thus avoid relying on `nn.topo`, since for 1. the `nn` attribute is not the same for each node).
Another solution might be to add a function `merge_networks networks` that would merge the roots and the topological orders of the networks. That could replace the line [`let nn = get_network input_node.(0)`](https://github.com/owlbarn/owl/blob/master/src/base/neural/owl_neural_graph.ml#L502) for layers with more than one input.

However I am not yet able to say what would need to be added to support training as well for one of these solutions, or if much more needs to be changed.

**Comment by ryanrhymes** [comment]
For 1, I think the most straightforward solution is introducing another function e.g. `inputs` which takes an array of shapes then returns an array of input nodes, and all these nodes are associated with the same network. Do you think this will address the issue?

For 2, similarly, we can make another function that takes an array of neuron nodes and outputs the network, e.g. `outputs` (if this makes sense, we probably even change `get_network` to `output` accordingly in future).

For 1, I don't think we need to do any changes in training. For 2, I guess not but I am not 100% sure atm. However, introducing new functions to address the issue is a safer way before we fully understand the problem. 

Let me know if I misunderstand anything. Also would you like to add these two functions? If not, I can certainly do that but probably only in the following days.

**Comment by pvdhove** [comment]
For 1. what you propose is indeed easier than what I wrote! I had assumed that the inputs could be created at different moments and thus multiple networks would have to be merged, but it is reasonable to suppose that the user knows the number of inputs at the beginning. For 2. it should be OK in inference mode (just collect the outputs once everything is run); in training mode, some adjustments might be needed (for instance in Keras it is possible to specify a different loss function and a weight for each output).
I'll give it a try to see if everything works easily or not :)
